File: StriverA2Z/JavaScript&TypeScript/src/Arrays/Medium/spiralPrinting.ts
function spiralPrinting(matrix: number[][]) {
  let n = matrix.length;
  let m = matrix[0].length;

  let left = 0;
  let top = 0;
  let right = m - 1;
  let bottom = n - 1;

  const ans = [];

  for (let i of matrix) {
    console.log(i);
  }

  while (top <= bottom && left <= right) {
    for (let i = left; i <= right; i++) {
      ans.push(matrix[top][i]);
    }
    top++;
    for (let i = top; i <= bottom; i++) {
      ans.push(matrix[i][right]);
    }
    right--;
    if (top <= bottom) {
      for (let i = right; i >= left; i--) {
        ans.push(matrix[bottom][i]);
      }
      bottom--;
    }
    if (left <= right) {
      for (let i = bottom; i >= top; i--) {
        ans.push(matrix[i][left]);
      }
      left++;
    }
  }

  console.log(ans);
}

spiralPrinting([
  [1, 2, 3],
  [8, 9, 4],
  [7, 6, 5],
]); // [1,2,3,6,9,8,7,4,5]

File: StriverA2Z/C++/3_ArraysQuestion/3.2_Medium/spiral_printing.cpp
#include<bits/stdc++.h>

using namespace std;

vector<int> spiralOrder(vector<vector<int>> matrix) {
    int n = matrix.size();
    int m = matrix[0].size();

    int left = 0;
    int top = 0;
    int right = m - 1;
    int bottom = n - 1;

    vector<int> ans;

    while (top <= bottom && left <= right) {
        for (int i = left; i <= right; i++) {
            ans.push_back(matrix[top][i]);
        }
        top++;
        for (int i = top; i <= bottom; i++) {
            ans.push_back(matrix[i][right]);
        }
        right--;
        if (top <= bottom) {
            for (int i = right; i >= left; i--) {
                ans.push_back(matrix[bottom][i]);
            }
            bottom--;
        }
        if (left <= right) {
            for (int i = bottom; i >= top; i--) {
                ans.push_back(matrix[i][left]);
            }
            left++;
        }
    }
    return ans;
}

int main() {
    vector<int> vec = spiralOrder({{1, 2, 3},
                                          {8, 9, 4},
                                          {7, 6, 5}});

    for (int x: vec) {
        cout << x << " ";
    }

    return 0;
}